ophyd/utils/__init__.py

```diff
@@ -75,3 +75,46 @@ def wrapper(f):
         return f
 
     return wrapper
+
+
+def _filtered_ip_ns():
+    import IPython
+    return {k: v
+            for k, v in IPython.get_ipython().user_ns.items()
+            if not v.startswith('_')}
+
+
+def instances_from_namespace(classes, *, ns=None):
+    '''Get instances of `classes` from the user namespace
+
+    Parameters
+    ----------
+    classes : type, or sequence of types
+        Passed directly to isinstance(), only instances of these classes
+        will be returned.
+
+    ns : Dict[str, Any], optional
+       namespace to pull from, defaults to getting the
+    '''
+    if ns is None:
+        ns = _filtered_ip_ns()
+    return [val for val in ns.values()
+            if isinstance(val, classes)]
+
+
+def ducks_from_namespace(attrs, *, ns=None):
+    '''Get instances that have all of attributes.
+
+    "Ducks" is a reference to "duck-typing." If it looks like a duck....
+
+    Parameters
+    ----------
+    attr : Union[str, Iterable[str]]
```

Comment:
I don't like that types are randomly creeping into the codebase without a plan. I think "string or iterable of strings" is better documentation for humans than `Union[str, Iterable[str]]`, at least today.

Reply:
I am resisting (on Steve's advice) putting them in as annotations, but I like having them in the docstrings as we could pragmatically extract them if want to.

It also gives us a consistent vocabulary which has been thought through carefully.

Reply:
Do you mean "programatically" or "pragmatically"? If "pragmatically", then I would say just stash them in a comment. I think we should either adopt this or not adopt it; I don't like the randomness of it.

Reply:
'programitically'.

I don't think it is worth going through and changing existing docstrings to use typing-style types, but is worth using for new docstrings and as we touch things.

Reply:
OK, I won't litigate this here.